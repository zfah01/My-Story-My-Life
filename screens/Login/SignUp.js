import React, { useState } from 'react';
import { Text, View, ScrollView, TouchableOpacity, TextInput, ImageBackground, Image } from 'react-native';
import Loading from '../../utils/Loading';
import { styles } from './Styles';
import { auth, db }from '../../firebase/firebase';



// Checks if an uppercase, lowercase, number, special character is contained in the password 
export const passwordValid = (str) => {
    const passwordStrong = /^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{8,}$/;
    return passwordStrong.test(str);
};

 //Check if the email is in a correct format 
export const emailValid = (str) => {
    const validEmail = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return validEmail.test(str);
};

export default function SignUp({ navigation }) {


    const [name, setName] = useState('');
    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const [confirmPassword, setConfirmPassword] = useState('');
    const [loading, setLoading] = useState(false);
    const [passwordCheck, setPasswordCheck] = useState(false);

    const RegisterToApp = () => {
        setLoading(true);
        
        if (!emailValid(email)) {
            setLoading(false);
            alert('Please ensure your email is valid.');
            return;
         }

        if (!passwordValid(password)) {
            setLoading(false);
            setPasswordCheck(true);
            alert('Please ensure your password is 8 characters long and contains: \n One uppercase letter, a number and a special character (#?!@$%^&*-)');
            return;
         }

        if (password !== confirmPassword) {
            setLoading(false);
            alert('Please check your passwords are entered correctly as they dont match.');
            return;
        }
        // Use createUserWithEmailAndPassword API to create a new account 
        auth.
            createUserWithEmailAndPassword(email, password)
            .then((response) => {
                const uid = response.user.uid;
                const data = {
                    id: uid,
                    email,
                    name,
                };
                
                // if user account is created successfully their data is stored to firebase
                // and is directed to HomeScreen 

                const usersRef = db.collection('users');
                usersRef
                    .doc(uid)
                    .set(data)
                    .then(() => {
                        setLoading(false);
                        navigation.navigate('HomeScreen');
                    })
                    .catch((error) => {
                        setLoading(false);
                        alert('Error: ' + error);
                    });
            })
            .catch((error) => {
                setLoading(false);
                alert('Error: ' + error);
            });
    };


   
        return (
                <View style={{backgroundColor: '#AFEEEE', width: '100%', height: '100%'}}>
                <View style={styles.intro}>
                <Text style={styles.heading}>Welcome to My Story, My Life</Text>
                <Text>Register for an account below:</Text>
            </View><ScrollView>
                    <View style={styles.mainView}>
                        <Image style={styles.logo} source={require('../../assets/logo.png')} />

                        <TextInput
                            style={styles.textInput}
                            placeholder='enter your full name...'
                            placeholderTextColor='#aaaaaa'
                            onChangeText={(textName) => setName(textName)}
                            value={name}
                            autoCapitalize='none'
                            testID='inputName'
                             />

                        <TextInput
                            style={styles.textInput}
                            placeholder='enter your email address...'
                            placeholderTextColor='#aaaaaa'
                            onChangeText={(textEmail) => setEmail(textEmail)}
                            value={email}
                            autoCapitalize='none'
                            testID='inputEmail'
                             />

                        <TextInput
                            style={styles.textInput}
                            secureTextEntry
                            placeholder='enter a password...'
                            placeholderTextColor='#aaaaaa'
                            onChangeText={(textPass) => setPassword(textPass)}
                            value={password}
                            autoCapitalize='none'
                            testID='inputPassword'
                           />

                        <TextInput
                            style={styles.textInput}
                            secureTextEntry
                            placeholder='confirm your password...'
                            placeholderTextColor='#aaaaaa'
                            onChangeText={(textPass) => setConfirmPassword(textPass)}
                            value={confirmPassword}
                            autoCapitalize='none'
                            testID='inputConfirmPassword'
                             />

                        <Text style={passwordCheck ? [styles.passwordGuidelines, { color: 'red' }] : styles.passwordGuidelines}>Password must be 8 or more characters and contain an upper case letter, numbers and a special character {'(#?!@$%^&*-)'}. </Text>

                        <TouchableOpacity
                            style={styles.buttonLogin}
                            onPress={() => RegisterToApp()}
                            testID='signUpButton'
                            accessibilityLabel='Create Account Button'>
                            <Text style={styles.loginText}> Create Account </Text>
                        </TouchableOpacity>

                        <View style={styles.footer}>
                            <Text style={styles.accountText}> Already have an account? <Text testID='loginLink'onPress={() => navigation.navigate('Welcome')} style={styles.createLink}> Log In</Text> </Text>
                        </View>
                    </View>
                </ScrollView>
                </View>
        );
    
}