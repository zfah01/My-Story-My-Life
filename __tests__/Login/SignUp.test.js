import 'react-native';
import React from 'react';
import SignUp, { passwordValid, emailValid } from '../../screens/Login/SignUp';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

import { cleanup, render, fireEvent } from '@testing-library/react-native';

afterEach(cleanup);

describe('Test to check Sign Up page and its functions', () => {

  const fakeNavigation = {
    navigate: jest.fn(),
  };

  it('Test to check Sign Up page is rendered', () => {
    const tree = renderer.create(<SignUp />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Greeting Message', () => {
    const title = 'Welcome to My Story, My Life';
    const { getByText } = render(<SignUp />); 

    const foundTitleText = getByText(title);
    expect(foundTitleText.props.children).toEqual(title);
  });

  it('Find button from  testID', () => {
    const {  getByTestId } = render(<SignUp />);
    const testIdName = 'signUpButton';
    const buttonID = getByTestId(testIdName);

    expect(buttonID).toBeTruthy();
  });

  it('Checks if text inputs recieve data.', () => {
    const { getByTestId } = render(<SignUp />);
    const idTestName = 'inputName';
    const idTestEmail= 'inputEmail';
    const idTestPassword = 'inputPassword';
    const inputName = getByTestId(idTestName);
    const inputEmail = getByTestId(idTestEmail);
    const inputPassword = getByTestId(idTestPassword);

    // Expect a blank entry when first loading the page.
    expect(inputName.props.value).toEqual('');
    expect(inputEmail.props.value).toEqual('');
    expect(inputPassword.props.value).toEqual('');

    // Add values to inputs.
    fireEvent.changeText(inputName, 'Stephen James');
    fireEvent.changeText(inputEmail, 'test@example.com');
    fireEvent.changeText(inputPassword, 'password');
    
    // Expect values to be stored in the inputs.
    expect(inputName.props.value).toEqual('Stephen James');
    expect(inputEmail.props.value).toEqual('test@example.com');
    expect(inputPassword.props.value).toEqual('password');
  });

  it('Checks that a password is valid', () => {
    const passwordBad = 'hello';
    const passwordGood = '@HelloHi1';
    expect(passwordValid(passwordBad)).toBeFalsy();
    expect(passwordValid(passwordGood)).toBeTruthy();
  });

  it('Checks that an email is valid', () => {
    const emailBad = 'hello';
    const emailGood = 'test@test.com';
    expect(emailValid(emailBad)).toBeFalsy();
    expect(emailValid(emailGood)).toBeTruthy();
  });

  it('Test Welcome page navigation', () => {
    const idLogin = 'loginLink';
    const { getByTestId } = render(<SignUp navigation={fakeNavigation} />);

    fireEvent(getByTestId(idLogin), 'press');
    expect(fakeNavigation.navigate).toBeCalledWith('Welcome');
  });

});