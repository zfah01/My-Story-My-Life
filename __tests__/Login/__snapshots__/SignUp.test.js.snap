// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to check Sign Up page and its functions Test to check Sign Up page is rendered 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "paddingTop": 70,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 25,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Welcome to My Story, My Life
    </Text>
    <Text>
      Register for an account below:
    </Text>
  </View>,
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 25,
            "justifyContent": "center",
            "margin": 20,
            "padding": 10,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "alignSelf": "center",
              "height": 100,
              "marginBottom": 10,
              "marginTop": 20,
              "width": 134,
            }
          }
        />
        <TextInput
          autoCapitalize="none"
          onChangeText={[Function]}
          placeholder="enter your full name..."
          placeholderTextColor="#aaaaaa"
          style={
            Object {
              "backgroundColor": "#f0efed",
              "borderRadius": 5,
              "height": 50,
              "marginBottom": 10,
              "marginLeft": 20,
              "marginRight": 20,
              "marginTop": 10,
              "overflow": "hidden",
              "paddingLeft": 10,
            }
          }
          testID="inputName"
          value=""
        />
        <TextInput
          autoCapitalize="none"
          onChangeText={[Function]}
          placeholder="enter your email address..."
          placeholderTextColor="#aaaaaa"
          style={
            Object {
              "backgroundColor": "#f0efed",
              "borderRadius": 5,
              "height": 50,
              "marginBottom": 10,
              "marginLeft": 20,
              "marginRight": 20,
              "marginTop": 10,
              "overflow": "hidden",
              "paddingLeft": 10,
            }
          }
          testID="inputEmail"
          value=""
        />
        <TextInput
          autoCapitalize="none"
          onChangeText={[Function]}
          placeholder="enter a password..."
          placeholderTextColor="#aaaaaa"
          secureTextEntry={true}
          style={
            Object {
              "backgroundColor": "#f0efed",
              "borderRadius": 5,
              "height": 50,
              "marginBottom": 10,
              "marginLeft": 20,
              "marginRight": 20,
              "marginTop": 10,
              "overflow": "hidden",
              "paddingLeft": 10,
            }
          }
          testID="inputPassword"
          value=""
        />
        <TextInput
          autoCapitalize="none"
          onChangeText={[Function]}
          placeholder="confirm your password..."
          placeholderTextColor="#aaaaaa"
          secureTextEntry={true}
          style={
            Object {
              "backgroundColor": "#f0efed",
              "borderRadius": 5,
              "height": 50,
              "marginBottom": 10,
              "marginLeft": 20,
              "marginRight": 20,
              "marginTop": 10,
              "overflow": "hidden",
              "paddingLeft": 10,
            }
          }
          testID="inputConfirmPassword"
          value=""
        />
        <Text
          style={
            Object {
              "color": "#8f8f8f",
              "fontSize": 14,
              "fontStyle": "italic",
              "textAlign": "center",
            }
          }
        >
          Password must be 8 or more characters and contain an upper case letter, numbers and a special character 
          (#?!@$%^&*-)
          . 
        </Text>
        <View
          accessibilityLabel="Create Account Button"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "#00e676",
              "borderRadius": 15,
              "justifyContent": "center",
              "marginTop": 20,
              "opacity": 1,
              "padding": 15,
              "width": 150,
            }
          }
          testID="signUpButton"
        >
          <Text
            style={
              Object {
                "color": "#000000",
                "fontSize": 18,
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
             Create Account 
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "marginTop": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 16,
              }
            }
          >
             Already have an account? 
            <Text
              onPress={[Function]}
              style={
                Object {
                  "color": "#6db8bf",
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
              testID="loginLink"
            >
               Log In
            </Text>
             
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;
